frame/tx-pause/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+//! - `unpause` - Unpause a pallet or optionally a specific call within a pallet.
+//!
+//! ## Usage
+//!
+//! The following examples show how to use the Transaction Pause pallet in your custom pallet.
+//! TODO check doc links
+//! ### Example within a runtime's [`pallet-frame-system::BaseCallFilter`] configuration:
+//!
+//! ```ignore
+//! impl frame_system::Config for Runtime {
+//! 	…
+//! 	type BaseCallFilter = InsideBoth<DefaultFilter, InsideBoth<TxPause, SafeMode>>;
+//! 	…
+//! }
+//! ```
+//!
+//! ## Genesis config
+//!
+//! The Transaction Pause pallet may be configured to pause calls on genesis with it's
+//! [`GenesisConfig`].
+//!
+//! ## Assumptions
+//!
+//! * TODO
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarking;
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+use frame_support::{
+	dispatch::GetDispatchInfo,
+	pallet_prelude::*,
+	traits::{CallMetadata, Contains, GetCallMetadata, IsSubType, IsType},
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::{traits::Dispatchable, DispatchResult};
+use sp_std::{convert::TryInto, prelude::*};
+
+pub use pallet::*;
+pub use weights::*;
+
+/// The stringy name of a pallet from [`GetCallMetadata`] for [`Config::RuntimeCall`] variants.
+pub type PalletNameOf<T> = BoundedVec<u8, <T as Config>::MaxNameLen>;
+/// The stringy name of a call (within a pallet) from [`GetCallMetadata`] for
+/// [`Config::RuntimeCall`] variants.
+pub type CallNameOf<T> = BoundedVec<u8, <T as Config>::MaxNameLen>;
+/// A sully specified pallet ([`PalletNameOf`]) and optional call ([`CallNameOf`])
+/// to partially or fully specify an item a variant of a  [`Config::RuntimeCall`].
+pub type FullNameOf<T> = (PalletNameOf<T>, Option<CallNameOf<T>>);
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ Dispatchable<Origin = Self::Origin>
+			+ GetDispatchInfo
+			+ GetCallMetadata
+			+ From<frame_system::Call<Self>>
+			+ IsSubType<Call<Self>>
+			+ IsType<<Self as frame_system::Config>::RuntimeCall>;
+
+		/// The only origin that can pause calls.
+		type PauseOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The only origin that can un-pause calls.
+		type UnpauseOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Contains all calls that cannot be paused.
+		///
+		/// The `TxMode` pallet cannot pause it's own calls, and does not need to be explicitly
+		/// added here.
+		type WhitelistCallNames: Contains<FullNameOf<Self>>;
+
+		/// Maximum length for pallet and call SCALE encoded string names.
+		///
+		/// Too long names will not be truncated but handled like
```

Comment:
We should add a test in the `integrity_test` hook to check that `AllPalletsWithSystem` do not overflow this.  
This would prevent parachains from misconfiguring it.